src/main/java/org/opensearch/agent/tools/RAGTool.java

```diff
@@ -0,0 +1,275 @@
+                vectorDBToolOutput = (T) "";
+            } else {
+                Gson gson = new Gson();
+                String[] hits = r.toString().split("\n");
+
+                StringBuilder resultBuilder = new StringBuilder();
+                for (String hit : hits) {
+                    JsonObject jsonObject = gson.fromJson(hit, JsonObject.class);
+                    String id = jsonObject.get("_id").getAsString();
+                    JsonObject source = jsonObject.getAsJsonObject("_source");
+
+                    resultBuilder.append("_id: ").append(id).append("\n");
+                    resultBuilder.append("_source: ").append(source.toString()).append("\n");
+                }
+
+                vectorDBToolOutput = (T) gson.toJson(resultBuilder.toString());
+            }
+
+            Map<String, String> tmpParameters = new HashMap<>();
+            tmpParameters.putAll(parameters);
+
+            if (vectorDBToolOutput instanceof List
+                && !((List) vectorDBToolOutput).isEmpty()
+                && ((List) vectorDBToolOutput).get(0) instanceof ModelTensors) {
+                ModelTensors tensors = (ModelTensors) ((List) vectorDBToolOutput).get(0);
+                Object response = tensors.getMlModelTensors().get(0).getDataAsMap().get("response");
+                tmpParameters.put(OUTPUT_FIELD, response + "");
+            } else if (vectorDBToolOutput instanceof ModelTensor) {
+                tmpParameters.put(OUTPUT_FIELD, escapeJson(toJson(((ModelTensor) vectorDBToolOutput).getDataAsMap())));
+            } else {
+                if (vectorDBToolOutput instanceof String) {
+                    tmpParameters.put(OUTPUT_FIELD, (String) vectorDBToolOutput);
+                } else {
+                    tmpParameters.put(OUTPUT_FIELD, escapeJson(toJson(vectorDBToolOutput.toString())));
+                }
+            }
+
+            RemoteInferenceInputDataSet inputDataSet = RemoteInferenceInputDataSet.builder().parameters(tmpParameters).build();
+            MLInput mlInput = MLInput.builder().algorithm(FunctionName.REMOTE).inputDataset(inputDataSet).build();
+            ActionRequest request = new MLPredictionTaskRequest(inferenceModelId, mlInput, null);
+
+            client.execute(MLPredictionTaskAction.INSTANCE, request, ActionListener.wrap(resp -> {
+                ModelTensorOutput modelTensorOutput = (ModelTensorOutput) resp.getOutput();
+                modelTensorOutput.getMlModelOutputs();
+                if (outputParser == null) {
+                    listener.onResponse((T) modelTensorOutput.getMlModelOutputs());
+                } else {
+                    listener.onResponse((T) outputParser.parse(modelTensorOutput.getMlModelOutputs()));
+                }
+            }, e -> {
+                log.error("Failed to run model " + inferenceModelId, e);
+                listener.onFailure(e);
+            }));
+        }, e -> {
+            log.error("Failed to search index.", e);
```

Comment:
Isn't the error here related to failure in running vectorDB tool than in searching the index? Maybe we can change the message?

Reply:
good question, it's related to `vectorDBTool.run`, but within `vectorDBTool.run`,  the error is mainly from `client.search`, because it's a recursive call.